runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,399 @@
+    super(name);
+    this.groupByKeyTransform = groupByKeyTransform;
+    this.fn = fn;
+  }
+
+  @Override
+  public PCollection<OutputT> apply(PCollection<InputT> input) {
+    PCollection.IsBounded isFnBounded =
+        DoFnSignatures.INSTANCE.getOrParseSignature(fn.getClass()).isBounded();
+    Coder<RestrictionT> restrictionCoder =
+        DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn).invokeGetRestrictionCoder();
+    KvCoder<InputT, RestrictionT> splitCoder = KvCoder.of(input.getCoder(), restrictionCoder);
+
+    return input
+        .apply(
+            "Pair with initial restriction",
+            ParDo.of(new PairWithRestrictionFn<InputT, OutputT, RestrictionT>(fn)))
+        .setCoder(splitCoder)
+        .apply("Split restriction", ParDo.of(new SplitRestrictionFn<InputT, RestrictionT>(fn)))
+        .apply(
+            "Assign unique key", ParDo.of(new AssignRandomUniqueKeyFn<KV<InputT, RestrictionT>>()))
+        .apply("Group by key", groupByKeyTransform.forInputCoder(splitCoder))
+        .apply(
+            "Process",
+            ParDo.of(
+                new ProcessFn<InputT, RestrictionT, OutputT, TrackerT>(
+                    fn,
+                    input.getCoder(),
+                    input.getWindowingStrategy().getWindowFn().windowCoder())))
+        .setIsBoundedInternal(isFnBounded);
+  }
+
+  /**
+   * Assigns a random unique key to each element of the input collection, so that the output
+   * collection is effectively the same elements as input, but with access to per-element (per-key)
+   * state and timers.
+   */
+  private static class AssignRandomUniqueKeyFn<T> extends DoFn<T, KV<String, T>> {
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      c.output(KV.of(UUID.randomUUID().toString(), c.element()));
+    }
+  }
+
+  /**
+   * Pairs each input element with its initial restriction using the given splittable {@link DoFn}.
+   */
+  private static class PairWithRestrictionFn<InputT, OutputT, RestrictionT>
+      extends DoFn<InputT, KV<InputT, RestrictionT>> {
+    private DoFn<InputT, OutputT> fn;
+    private transient DoFnInvoker<InputT, OutputT> invoker;
+
+    PairWithRestrictionFn(DoFn<InputT, OutputT> fn) {
+      this.fn = fn;
+    }
+
+    @Setup
+    public void setup() {
+      invoker = DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn);
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      context.output(
+          KV.of(
+              context.element(),
+              invoker.<RestrictionT>invokeGetInitialRestriction(context.element())));
+    }
+  }
+
+  /**
+   * The heart of splittable {@link DoFn} execution: processes a single (element, restriction) pair
+   * by creating a tracker for the restriction and checkpointing/resuming processing later if
+   * necessary.
+   *
+   * <p>TODO: This uses deprecated OldDoFn since DoFn does not provide access to state/timer
```

Comment:
To enable dynamic work rebalancing and give runners more control into the execution of a split, this seems as though it might be better to rewrite in a manner similar to `ReduceFnRunner` that exercises the underlying APIs directly, but exposes additional handles to enable the runner to more richly interact with the abstractions (especially `RestrictionTracker` similar to the current `Reader/Source` APIs)